x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/checkpoint/DataFrameTransformsCheckpointService.java

```diff
@@ -96,6 +117,105 @@ public void getCheckpoint(DataFrameTransformConfig transformConfig, long checkpo
+
+        CountDownLatch latch = new CountDownLatch(numberOfOperations);
+        Checkpoints checkpoints = new Checkpoints();
+        AtomicBoolean failed = new AtomicBoolean(false);
+
+        // get the current checkpoint
+        if (current != 0) {
+            dataFrameTransformsConfigManager.getTransformCheckpoint(task.getTransformId(), current,
+                    new LatchedActionListener<>(ActionListener.wrap(checkpoint -> checkpoints.currentCheckpoint = checkpoint, e -> {
+                        logger.debug("Failed to retrieve checkpoint [" + current + "] for data frame []" + task.getTransformId(), e);
+                        // only call onFailure if not already happened
+                        if (failed.compareAndSet(false, true)) {
+                            listener.onFailure(new CheckpointException("Failed to retrieve current checkpoint [" + current + "]", e));
+                        }
+                    }), latch));
+        }
+
+        // get the in-progress checkpoint
+        if (inProgress != 0) {
+            dataFrameTransformsConfigManager.getTransformCheckpoint(task.getTransformId(), inProgress,
+                    new LatchedActionListener<>(ActionListener.wrap(checkpoint -> checkpoints.inProgressCheckpoint = checkpoint, e -> {
+                        logger.debug("Failed to retrieve in progress checkpoint [" + inProgress + "] for data frame ["
+                                + task.getTransformId() + "]", e);
+                        // only call onFailure if not already happened
+                        if (failed.compareAndSet(false, true)) {
+                            listener.onFailure(
+                                    new CheckpointException("Failed to retrieve in progress checkpoint [" + inProgress + "]", e));
+                        }
+                    }), latch));
+        }
+
+        // get the current state
+        dataFrameTransformsConfigManager.getTransformConfiguration(task.getTransformId(), ActionListener.wrap(transformConfig -> {
+            getCheckpoint(transformConfig,
+                    new LatchedActionListener<>(ActionListener.wrap(checkpoint -> checkpoints.sourceCheckpoint = checkpoint, e2 -> {
+                        logger.debug("Failed to retrieve actual checkpoint for data frame [" + task.getTransformId() + "]", e2);
+
+                        // only call onFailure if not already happened
+                        if (failed.compareAndSet(false, true)) {
+                            listener.onFailure(
+                                    new CheckpointException("Failed to retrieve actual checkpoint", e2));
+                        }
+                    }), latch));
+
+        }, e -> {
+            logger.warn("Failed to retrieve configuration for data frame [" + task.getTransformId() + "]", e);
+
+            // only call onFailure if not already happened
+            if (failed.compareAndSet(false, true)) {
```

Comment:
This is error prone. Why not simply do a callback system like we usually do? I think a stack of `ActionListener`s and each calling back the other makes the most sense. I suppose some branches will have to be done internally like
```
if (inProgress != 0) {
   dataFrameTransformsConfigManager.getTransformCheckpoint(task.getTransformId(), inProgress, inProgressListener);
} else {
   inProgressListener.onResponse(null);
}
```
But that will make it more clear I think, and less state to worry about. 

I am also not 100% the behavior of doing `latch.await` on this thread when it was previously async up the stack. I think that still pauses the current thread execution, but it only awakens it when the latch gets signaled, right? It is definitely not the same as a spin wait where the thread is not able to be rotated off from being currently executed.

I should probably read Java Concurrency in Practice :/

Reply:
I like to avoid sequential execution as it isn't necessary here and makes it unnecessary slow.

I admit, the code is complicated, especially the code that calculates the countdown value upfront. It can be simplified by using always `3` parallel task while some might be no-ops like in your example.

The error handling is still tricky, the code must ensure that the listener is only called exactly once, but there is a `NotifyOnceListener` I should use.

As for the latch: yes afaik that's the usecase they describe and it's build with conditionals, so the await thread does not block but goes to sleep and lets others do their business. Once the latch reaches `0` it gets a signal.

Again, I understand this is way to complicated, but with proper simplifications, breakdowns and using existing abstractions it should get better.